plutus-core/plutus-core/src/Data/SatInt.hs

```diff
@@ -0,0 +1,297 @@
+{- |
+Adapted from 'Data.SafeInt' to perform saturating arithmetic (i.e. returning max or min bounds) instead of throwing on overflow.
+
+This is not quite as fast as using 'Int' or 'Int64' directly, but we need the safety.
```

Comment:
What makes `SatInt` safer than `Int`? If we are sure that no overflow can possibly occur, then they are equally safe. If we're not, then I'd argue that `Int` is safer, because it crashes the program instead of silently allowing to run the program at a cheaper price than what it actually costs. An exception is something that we can easily catch. But with `SatInt` we don't even know if something has gone wrong. This seems very dangerous to me. Am I missing something?

Reply:
`Int` -> overflow just overflows, i.e. wraps
`SafeInt` (on which `SatInt` is based) -> throws on overflow
`SatInt` -> saturates on overflow

Maybe we should use `SafeInt` indeed! However, the only case in which `SatInt` allows someone to run a program more cheaply than they should is where the budget is `maxBound`! In any other case, it will saturate to a value which is smaller than the true amount... but still larger than the budget. So if we were really worried we could insist that our budgets were less than `maxBound`.

I'm not sure that no overflow can occur, indeed I think it probably can as part of an attack. What I *was* more sure about was that no *budget* would be above `maxBound`. As I argued above, it doesn't matter if the calculation for the amount spent overflows, so long as it remains larger than the budget, which is what saturating operations give us.

Reply:
> However, the only case in which `SatInt` allows someone to run a program more cheaply than they should is where the budget is `maxBound`!

I've finally got it, thank you.

Reply:
I redid the comment to include all this too.